### ISSUE TITLE
react/jsx-no-bind false positive?

### ISSUE THREAD

**Issue by bcutler-work** [yannickcr/eslint-plugin-react#1417]
Example snippet:

```jsx
import React from 'react';
import PropTypes from 'prop-types';

class FalsePositive extends React.PureComponent {
  static propTypes = {
    list: PropTypes.arrayOf(PropTypes.node).isRequired,
  };

  render() {
    return (
      <div>
        {this.props.list.map(this.wrap.bind(this, 'span'))}
      </div>
    );
  }

  wrap(Tag, item, i) {
    return <Tag key={i}>{item}</Tag>;
  }
}

export default FalsePositive;
```

`this.wrap.bind` triggers the rule, even though the bound function isn't the actual prop being passed, but is being used during the render to build the children, so it shouldn't break purity.

**Comment by jackyho112** [comment]
I will give this one a look. 😄

**Comment by jackyho112** [comment]
@ljharb 

Seems like the rule needs a bit more fixing than I thought it would to only check for `bind` in JSX props. If that was its purpose, [this test](https://github.com/yannickcr/eslint-plugin-react/blob/master/tests/lib/rules/jsx-no-bind.js#L168) and [this one](https://github.com/yannickcr/eslint-plugin-react/blob/master/tests/lib/rules/jsx-no-bind.js#L168) should not warn but they are currently invalid tests.

The [current logic](https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/jsx-no-bind.js#L47) in the rule checks for all `bind` calls in all JSX methods and warn each of them. And along with the [documentation](https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/jsx-no-bind.js#L18) in that file, I am given the impression that this rule does check for `bind` calls in a JSX method instead of only those in JSX attributes.

**Comment by ljharb** [comment]
The rule is to avoid new functions being created *and passed as props to an element* (it should really apply not just to jsx, but to createElement calls as well).

**Comment by ljharb** [comment]
It's hard to evaluate without an actual PR; I'd love to look at the test case and docs changes in particular that you're suggesting.

**Comment by jackyho112** [comment]
@ljharb 

Thanks for replying! I think I know the direction I'm going with to fix this issue now. Will open a PR soon! 😄
